addons/mail/static/src/models/audio.js

```diff
@@ -0,0 +1,118 @@
+         * @param {float} [param0.volume] the volume percentage in decimal to play this sound.
+         *   If not provided, uses the default volume of this sound effect.
+         */
+        play({ loop = false, volume } = {}) {
+            if (this.messaging.isInQUnitTest) {
+                return;
+            }
+            if (!this.audio) {
+                return;
+            }
+            this.audio.pause();
+            this.audio.currentTime = 0;
+            this.audio.loop = loop;
+            this.audio.volume = volume !== undefined ? volume : this.defaultVolume;
+            Promise.resolve(this.audio.play()).catch(e => this._onAudioPlayError(e));
+        },
+        /**
+         * Resets the audio to the start of the track and pauses it.
+         */
+        stop() {
+            if (!this.audio) {
+                return;
+            }
+            this.audio.pause();
+            this.audio.currentTime = 0;
+        },
+        /**
+         * @private
+         * @param {DOMException} error
+         */
+        _onAudioPlayError(error) {
+            this.update({ audioPlayError: error });
+            // error on play can trigger fallback to .mp3; retry in case this
+            // solved the problem.
+            this.audio.play().catch();
+        },
+        /**
+         * @private
+         */
+        _onSourceChanged() {
+            if (!this.audio) {
+                return;
+            }
+            this.audio.src = this.source;
+        },
+    },
+    fields: {
+        /**
+         * HTMLAudioElement
+         * Does not require to be mounted on the DOM to operate.
+         */
+        audio: attr(),
+        audioPlayError: attr(),
+        /**
+         * The default volume to play this sound effect, when unspecified.
+         */
+        defaultVolume: attr({
+            default: 1,
+        }),
+        extension: attr({
+            compute() {
+                if (!this.audio) {
+                    return clear();
+                }
+                // If the device has tried to play audio and failed, perhaps ogg
+                // is not supported -> fallback to mp3.
+                if (this.audioPlayError && this.audioPlayError.name === 'NotSupportedError') {
+                    return '.mp3';
+                }
+                return this.audio.canPlayType('audio/ogg; codecs=vorbis') ? '.ogg' : '.mp3';
+            },
+        }),
+        /**
+         * Name of the audio file.
+         */
+        filename: attr({
+            identifying: true,
+        }),
+        /**
+         * Path to the audio file.
+         */
+        path: attr({
+            default: '/mail/static/src/audio/',
+            identifying: true,
+        }),
+        source: attr({
+            compute() {
+                return `${this.path}${this.filename}${this.extension}`;
+            },
```

Comment:
Note that `extension` can become `undefined`, so `source` can have "undefined" at the end.

If models framework did not enforce ordering of execution of `onChange` and computes, and `audio` was not coupled to model, then could theoretically have `WebAudio/src` having this strange source path with `undefined` at the end, which itself could (silently, fail fetch of resource) crash.

But it doesn't practically happen thanks to implementation detail of models framework.